File: factors.go
package main

import (
	"fmt"
	"math/big"
)

// A hash map to store prime numbers for easy look up
var resultsCache map[string]string

func init() {
	// initialize the hash map
	resultsCache = make(map[string]string)
}

// factorizes as many numbers as possible into a product
// of two smaller numbers and prints the result
func printPrimeFactors(number *big.Int, oddPrime *big.Int) int {
	zero := big.NewInt(0)
	one := big.NewInt(1)
	two := big.NewInt(2)
	limit := big.NewInt(611953)
	stopLoop := 20000000

	// now I don't care what really happens, let us go until we can't
	if oddPrime.Cmp(limit) >= 0 {
		limit.Set(big.NewInt(30597650))
	} else {
		stopLoop = 300000000
	}

	// Handle the case when the number is less than or equal to 1
	if number.Cmp(one) <= 0 {
		return 0
	}

	// check for cached values
	cachedResult, found := resultsCache[number.String()]
	if found {
		fmt.Printf("%s=%s\n", number.String(), cachedResult)
		return 0
	}

	// Handle the case when the number is divisible by 2
	if new(big.Int).Mod(number, two).Cmp(zero) == 0 {
		quotient := new(big.Int)
		quotient.Quo(number, two)
		resultsCache[number.String()] = quotient.String() + "*2"

		// print the result and exit from here
		fmt.Printf("%s=%s*2\n", number.String(), quotient.String())
		return 0
	}

	sqrtNumber := new(big.Int).Set(number)
	sqrtNumber.Sqrt(sqrtNumber)

	loopCounter := 0
	for oddPrime.Cmp(sqrtNumber) <= 0 {
		if new(big.Int).Mod(number, oddPrime).Cmp(zero) == 0 {
			quotient := new(big.Int)

			// this is an expensive operation so we'd save the result for later
			quotient.Quo(number, oddPrime)

			// let's save the result for later
			resultsCache[number.String()] = quotient.String() + "*" +
				oddPrime.String()
			fmt.Printf("%s=%s*%s\n", number.String(), quotient.String(),
				oddPrime.String())
			return 0
		}
		// skip this number if we go past this prime number without a match
		if oddPrime.Cmp(limit) > 0 || loopCounter > stopLoop {
			// fmt.Printf("Limit reached for [%s]\n", number.String())
			return 1
		}
		oddPrime.Add(oddPrime, two) // oddPrime += 2
		loopCounter++
	}

	// the number is a prime
	fmt.Printf("%s=%s*1\n", number.String(), number.String())
	return 0
}

File: factors.c
#include </usr/local/include/gmp.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>

void print_prime_factors(mpz_t number);

/**
 * main - Prime factorization
 * @argc: argument counter
 * @argv: argument vector
 *
 * Return: 0 on success, else non-zero
 */
int main(int argc, char *argv[])
{
	FILE *file;
	char *buffer = NULL;
	size_t n = 0;
	mpz_t number;

	if (argc != 2)
	{
		fprintf(stderr, "USAGE: factors file\n");
		exit(EXIT_FAILURE);
	}

	file = fopen(argv[1], "r");
	if (file == NULL)
	{
		fprintf(stderr, "Error: Can't open file %s\n", argv[1]);
		exit(EXIT_FAILURE);
	}

	while (getline(&buffer, &n, file) != -1)
	{
		/* initialize the number and check for errors */
		if (mpz_init_set_str(number, buffer, 10) == -1)
		{
			mpz_clear(number);
			continue; /* yes, it was invalid but just skip it */
		}
		print_prime_factors(number);
	}

	free(buffer);
	fclose(file);

	mpz_clear(number);

	if (errno == ENOMEM)
	{
		fprintf(stderr, "Error: malloc failed\n");
		exit(EXIT_FAILURE);
	}

	return (EXIT_SUCCESS);
}

/**
 * print_prime_factors - factorizes as many numbers as possible into a product
 * of two smaller numbers and prints the result
 * @number: the number to factorize
 */
void print_prime_factors(mpz_t number)
{
	mpz_t odd_prime, quotient;

	mpz_init(odd_prime);
	mpz_init(quotient);

	if (mpz_cmp_si(number, 1) <= 0)
		return; /* skip 1 and anything below it */

	/* let's check whether it's divisible by 2 and perform an early return */
	if (mpz_even_p(number) != 0)
	{
		mpz_divexact_ui(quotient, number, 2);
		gmp_printf("%Zd=%Zd*%d\n", number, quotient, 2);
		return;
	}

	/* let's use odd numbers and step through the number, it wasn't even */
	mpz_set_ui(odd_prime, 3);
	mpz_sqrt(quotient, number);

	while (mpz_cmp(odd_prime, quotient) <= 0)
	{
		if (mpz_divisible_p(number, odd_prime) != 0)
		{
			mpz_divexact(quotient, number, odd_prime);
			/* print the result*/
			gmp_printf("%Zd=%Zd*%Zd\n", number, quotient, odd_prime);

			/* clean up and return */
			mpz_clear(odd_prime);
			mpz_clear(quotient);
			return;
		}

		/* increment odd_prime by 2 */
		mpz_add_ui(odd_prime, odd_prime, 2); /* odd_prime += 2*/
	}

	/* the number is a prime number */
	gmp_printf("%Zd=%Zd*%d\n", number, number, 1);

	/* clean up what was used */
	mpz_clear(odd_prime);
	mpz_clear(quotient);
}
